DQM/SiPixelMonitorDigi/src/SiPixelDigiModule.cc

```diff
@@ -48,8 +51,7 @@ void SiPixelDigiModule::book(const edm::ParameterSet& iConfig,
                              bool additInfo,
                              bool isUpgrade) {
   //isUpgrade = iConfig.getUntrackedParameter<bool>("isUpgrade");
-  edm::ESHandle<TrackerTopology> tTopoHandle;
-  iSetup.get<TrackerTopologyRcd>().get(tTopoHandle);
+  edm::ESHandle<TrackerTopology> tTopoHandle = iSetup.getHandle(trackerTopoTokenBeginRun_);
```

Comment:
@makortel, the code is crashing on this line. Here is an output when running step 3 from workflow 4.53
```
A fatal system signal has occurred: abort signal
The following is the call stack containing the origin of the signal.

Mon Nov 23 17:25:06 CET 2020
Thread 5 (Thread 0x7f64be5ff700 (LWP 82214)):
#0  0x00007f6524a839f5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f652506762c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f64d476ec1c in Eigen::ThreadPoolTempl<tensorflow::thread::EigenEnvironment>::WorkerLoop(int) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#4  0x00007f64d476bb43 in std::_Function_handler<void (), tensorflow::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#5  0x00007f652506cd2f in std::execute_native_thread_routine (__p=0x7f64e233ef40) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#6  0x00007f6524a7fe65 in start_thread () from /lib64/libpthread.so.0
#7  0x00007f65247a888d in clone () from /lib64/libc.so.6
Thread 4 (Thread 0x7f64bf3ff700 (LWP 82213)):
#0  0x00007f6524a839f5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f652506762c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f64d476ec1c in Eigen::ThreadPoolTempl<tensorflow::thread::EigenEnvironment>::WorkerLoop(int) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#4  0x00007f64d476bb43 in std::_Function_handler<void (), tensorflow::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#5  0x00007f652506cd2f in std::execute_native_thread_routine (__p=0x7f64c70b8920) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#6  0x00007f6524a7fe65 in start_thread () from /lib64/libpthread.so.0
#7  0x00007f65247a888d in clone () from /lib64/libc.so.6
Thread 3 (Thread 0x7f64c0852700 (LWP 82212)):
#0  0x00007f6524a839f5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f652506762c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f64d476ec1c in Eigen::ThreadPoolTempl<tensorflow::thread::EigenEnvironment>::WorkerLoop(int) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#4  0x00007f64d476bb43 in std::_Function_handler<void (), tensorflow::thread::EigenEnvironment::CreateThread(std::function<void ()>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/external/slc7_amd64_gcc820/lib/libtensorflow_framework.so.2
#5  0x00007f652506cd2f in std::execute_native_thread_routine (__p=0x7f64c70b8800) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#6  0x00007f6524a7fe65 in start_thread () from /lib64/libpthread.so.0
#7  0x00007f65247a888d in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f64fa703700 (LWP 81994)):
#0  0x00007f6524a87199 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f6519dfffd7 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f6519e00a9a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f652506cd2f in std::execute_native_thread_routine (__p=0x7f651a0fa240) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00007f6524a7fe65 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f65247a888d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f6522d90540 (LWP 81761)):
#0  0x00007f652479dbed in poll () from /lib64/libc.so.6
#1  0x00007f6519e0043f in full_read.constprop () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f6519e00b7c in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f6519e01a59 in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00007f65246e0337 in raise () from /lib64/libc.so.6
#6  0x00007f65246e1a28 in abort () from /lib64/libc.so.6
#7  0x00007f65246d9156 in __assert_fail_base () from /lib64/libc.so.6
#8  0x00007f65246d9202 in __assert_fail () from /lib64/libc.so.6
#9  0x00007f64ca4aa959 in edm::eventsetup::EventSetupRecordImpl::getImplementation<TrackerTopology> (this=0x7f6491f56560, iData=@0x7ffcffa55c18: 0x0, iProxyIndex=..., iTransientAccessOnly=false, oDesc=@0x7ffcffa55c10: 0x0, whyFailedFactory=..., iEventSetupImpl=0x7f6494321580) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/FWCore/Framework/interface/EventSetupRecordImpl.h:207
#10 0x00007f64ca4a9953 in edm::eventsetup::EventSetupRecord::getHandleImpl<edm::ESHandle, TrackerTopology, TrackerTopologyRcd> (this=0x7ffcffa55cd0, iToken=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/FWCore/Framework/interface/EventSetupRecord.h:230
#11 0x00007f64ca4a89b3 in edm::eventsetup::EventSetupRecordImplementation<TrackerTopologyRcd>::getHandle<TrackerTopology> (this=0x7ffcffa55cd0, iToken=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/FWCore/Framework/interface/EventSetupRecordImplementation.h:49
#12 0x00007f64ca4a76e9 in edm::EventSetup::getHandle<TrackerTopology, TrackerTopologyRcd> (this=0x7ffcffa59d50, iToken=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/FWCore/Framework/interface/EventSetup.h:154
#13 0x00007f64ca49eb7c in SiPixelDigiModule::book (this=0x7f647fb7da00, iConfig=..., iSetup=..., iBooker=..., type=1, twoD=false, hiRes=false, reducedSet=true, additInfo=false, isUpgrade=false) at /home/ferencek/PixelOffline/CMSSW_11_2_0_pre8/src/DQM/SiPixelMonitorDigi/src/SiPixelDigiModule.cc:54
#14 0x00007f64ca4b8986 in SiPixelDigiSource::bookMEs (this=0x7f64e6e93940, iBooker=..., iSetup=...) at /home/ferencek/PixelOffline/CMSSW_11_2_0_pre8/src/DQM/SiPixelMonitorDigi/src/SiPixelDigiSource.cc:1272
#15 0x00007f64ca4b150b in SiPixelDigiSource::bookHistograms (this=0x7f64e6e93940, iBooker=..., iSetup=...) at /home/ferencek/PixelOffline/CMSSW_11_2_0_pre8/src/DQM/SiPixelMonitorDigi/src/SiPixelDigiSource.cc:314
#16 0x00007f64ca4ceded in DQMOneEDAnalyzer<edm::EndLuminosityBlockProducer, edm::one::WatchLuminosityBlocks>::beginRun(edm::Run const&, edm::EventSetup const&)::{lambda(dqm::implementation::IBooker&)#1}::operator()(dqm::implementation::IBooker&) const (this=0x7f64e6e93940, booker=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/DQMServices/Core/interface/DQMOneEDAnalyzer.h:44
#17 0x00007f64ca4cfad8 in dqm::implementation::DQMStore::bookTransaction<DQMOneEDAnalyzer<edm::EndLuminosityBlockProducer, edm::one::WatchLuminosityBlocks>::beginRun(edm::Run const&, edm::EventSetup const&)::{lambda(dqm::implementation::IBooker&)#1}>(DQMOneEDAnalyzer<edm::EndLuminosityBlockProducer, edm::one::WatchLuminosityBlocks>::beginRun(edm::Run const&, edm::EventSetup const&)::{lambda(dqm::implementation::IBooker&)#1}, unsigned long, bool) (this=0x7f65129c9c00, f=..., moduleId=187, canSaveByLumi=true) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/DQMServices/Core/interface/DQMStore.h:628
#18 0x00007f64ca4cef81 in DQMOneEDAnalyzer<edm::EndLuminosityBlockProducer, edm::one::WatchLuminosityBlocks>::beginRun (this=0x7f64e6e93940, run=..., setup=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/DQMServices/Core/interface/DQMOneEDAnalyzer.h:41
#19 0x00007f652733857b in edm::one::EDProducerBase::doBeginRun(edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#20 0x00007f6527316c50 in edm::WorkerT<edm::one::EDProducerBase>::implDoBegin(edm::RunTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#21 0x00007f652722e3d6 in decltype ({parm#1}()) edm::convertException::wrap<edm::Worker::runModule<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::Context const*)::{lambda()#1}>(edm::Worker::runModule<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::Context const*)::{lambda()#1}) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#22 0x00007f652722e5df in bool edm::Worker::runModule<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::Context const*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#23 0x00007f652722e836 in std::__exception_ptr::exception_ptr edm::Worker::runModuleAfterAsyncPrefetch<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >(std::__exception_ptr::exception_ptr const*, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0>::Context const*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#24 0x00007f652722ecc4 in void edm::SerialTaskQueueChain::actionToRun<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&>(edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#25 0x00007f652722ed81 in edm::SerialTaskQueue::QueuedTask<edm::SerialTaskQueueChain::push<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&>(edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::RunPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&)::{lambda()#1}>::execute() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#26 0x00007f6525a74bfd in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::process_bypass_loop (this=this@entry=0x7f65217ab200, context_guard=..., t=t@entry=0x7f63dab10b40, isolation=isolation@entry=0) at ../../src/tbb/custom_scheduler.h:393
#27 0x00007f6525a74ef5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7f65217ab200, parent=..., child=<optimized out>) at ../../include/tbb/task.h:1003
#28 0x00007f652720aa77 in edm::EventProcessor::beginRun(edm::Hash<2> const&, unsigned int, bool&, bool&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#29 0x00007f652720b3b8 in edm::EventProcessor::runToCompletion() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#30 0x000000000040f59d in tbb::interface7::internal::delegated_function<main::{lambda()#1}::operator()() const::{lambda()#1} const, void>::operator()() const ()
#31 0x00007f6525a6fbc1 in tbb::interface7::internal::task_arena_base::internal_execute (this=0x7ffcffa5a9d0, d=...) at ../../src/tbb/arena.cpp:1105
#32 0x0000000000410561 in main::{lambda()#1}::operator()() const ()
#33 0x000000000040eff5 in main ()

Current Modules:

Module: SiPixelDigiSource:SiPixelDigiSource (crashed)

A fatal system signal has occurred: abort signal
Aborted (core dumped)
```

However, I don's see anything obviosuly wrong. Any ideas? Thanks.

Reply:
The stack trace indicates an assertion failure. Could you paste the assertion failure printout as well (should be a few lines above the stack trace)?

Reply:
@makortel, here it is
```
cmsRun: /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre8/src/FWCore/Framework/interface/EventSetupRecordImpl.h:208: void edm::eventsetup::EventSetupRecordImpl::getImplementation(const DataT*&, edm::ESProxyIndex, bool, const edm::eventsetup::ComponentDescription*&, std::shared_ptr<edm::ESHandleExceptionFactory>&, const edm::EventSetupImpl*) const [with DataT = TrackerTopology]: Assertion `iProxyIndex.value() > -1 and iProxyIndex.value() < static_cast<ESProxyIndex::Value_t>(keysForProxies_.size())' failed.
```

Reply:
Thanks, although it didn't really help since your stack trace already had line numbers. I need to try this on my own, apparently any of the failed workflow above will suffice.

Reply:
@ferencek One problem is that `SiPixelDigiModule` is constructed in `SiPixelDigiSource::buildStructure()`
https://github.com/cms-sw/cmssw/blob/91c264fa0ec3e9cabbbaea6ea11a6e3cf6cf3514/DQM/SiPixelMonitorDigi/src/SiPixelDigiSource.cc#L1125
which is called by `SiPixelDigiSource::dqmBeginRun()`, which is too late for consumes registration. The `esConsumes()` calls must be done in the EDModule constructor (or in functions called from the constructor).

An attempt to register (es)consumes elsewhere should nevertheless not result an assertion failure but an explanatory exception. I see now that we do have such a check for ED-consumes, but not for ES-consumes. Thanks for the report!

Reply:
@makortel, thanks a lot for the clarification. I think I see a possible way to fix this which I will try to implement and test. I also see that the same issue also exists in a few other modules and I guess SiPixelDigiSource was the first one crashing in the job. I will have to fix those other modules as well.

Reply:
> I see now that we do have such a check for ED-consumes, but not for ES-consumes

This check is added in https://github.com/cms-sw/cmssw/pull/32270 (and I checked that the earlier version of this PR fired the exception of that PR).